ledger/simulation/simulator.go

```diff
@@ -0,0 +1,190 @@
+type ScopedSimulatorError struct {
+	SimulatorError        // the original error for internal use
+	External       string // the external error for public use
+}
+
+// ==============================
+// > Simulator
+// ==============================
+
+// Simulator is a transaction group simulator for the block evaluator.
+type Simulator struct {
+	ledger simulatorLedger
+}
+
+// MakeSimulator creates a new simulator from a ledger.
+func MakeSimulator(ledger *data.Ledger) *Simulator {
+	return &Simulator{
+		ledger: simulatorLedger{ledger, ledger.Latest()},
+	}
+}
+
+// check verifies that the transaction is well-formed and has valid or missing signatures.
+// An invalid transaction group error is returned if the transaction is not well-formed or there are invalid signatures.
+// To make things easier, we support submitting unsigned transactions and will respond whether signatures are missing.
+func (s Simulator) check(hdr bookkeeping.BlockHeader, txgroup []transactions.SignedTxn) (err error, isMissingSigs bool) {
+	specialAddresses := transactions.SpecialAddresses{
+		FeeSink:     hdr.FeeSink,
+		RewardsPool: hdr.RewardsPool,
+	}
+	consensusParams := config.Consensus[hdr.CurrentProtocol]
+
+	// Filter signed and unsigned transactions
+	var signedTxns []transactions.SignedTxn
+	for _, tx := range txgroup {
+		// If unsigned, check that they are well-formed
+		if verify.TxnIsMissingSig(&tx) {
+			isMissingSigs = true
+			if err = tx.Txn.WellFormed(specialAddresses, consensusParams); err != nil {
+				err = InvalidTxGroupError{SimulatorError{err}}
+				return
+			}
+			continue
+		}
+
+		// Otherwise add to the list of signed transactions
+		signedTxns = append(signedTxns, tx)
+	}
+
+	// Guard against no signed transactions
+	if len(signedTxns) == 0 {
+		return
+	}
+
+	// Verify the signed transactions are well-formed and have valid signatures
+	_, err = verify.TxnGroup(signedTxns, hdr, nil)
+	if err != nil {
+		err = InvalidTxGroupError{SimulatorError{err}}
+		return
+	}
+
+	return
+}
+
+func (s Simulator) evaluate(hdr bookkeeping.BlockHeader, stxns []transactions.SignedTxn) (*ledgercore.ValidatedBlock, error) {
+	// s.ledger has 'StartEvaluator' because *data.Ledger is embedded in the simulatorLedger
+	// and data.Ledger embeds *ledger.Ledger
+	eval, err := s.ledger.StartEvaluator(hdr, len(stxns), 0)
+	if err != nil {
+		return nil, err
+	}
+
+	group := transactions.WrapSignedTxnsWithAD(stxns)
+
+	err = eval.TransactionGroup(group)
+	if err != nil {
+		return nil, err
+	}
+
+	// Finally, process any pending end-of-block state changes.
+	vb, err := eval.GenerateBlock()
+	if err != nil {
+		return nil, err
+	}
+
+	return vb, nil
+}
+
+// Simulate simulates a transaction group using the simulator. Will error if the transaction group is not well-formed or an
+// unexpected error occurs. Otherwise, evaluation failure messages are returned.
+func (s Simulator) Simulate(txgroup []transactions.SignedTxn) (generated.SimulationResult, error) {
```

Comment:
I don't really like these `generated` types bleeding out of the `daemon/algod/api` backwater they were born in. But I suppose that might go away later.  I think you're going to need to return that `vb` from `s.evaluate` eventually, and pick it apart in the caller, rather than here.

Reply:
I wasn't fond of them bleeding out either; how is similar behavior handled elsewhere? Regarding where to pick apart `vb` in the future, I think it boils down to where/how we expect this `Simulate` method to be used.

If this method is only used by the simulation endpoint, then it seems reasonable to me for the picking apart and restructuring to live here.

But if we foresee it being used in more places than just the simulation endpoint, then I'd support changing this function declaration to `func (s Simulator) Simulate(txgroup []transactions.SignedTxn) (vb *ledgercore.ValidatedBlock, missingSignatures bool, evalFailure string, err error)` or some local, structured `SimulationResult` analog to avoid the bleeding out issue.

Reply:
I think we try to return "normal" types at least into the NodeInterface methods, or maybe all the way into handlers.go, and then we write the routines that pick them apart and make the new `generated` datastructures that the API offers.

I think it's reasonable to think that `Simulate` is something that various callers will need (some for debug, some to return values). I would not want to hamper such callers by forcing them to use the "weaker" abstraction, when the `vb` is probably what they want.

Reply:
That makes sense - I came to a similar conclusion while writing my previous comment. I'll make the `Simulate` methods more general and, in the future, support passing optional debugger hooks to them.